File: android_app/app/src/main/java/com/oceansentinels/app/mesh/ble/DeviceIdentifier.kt
package com.oceansentinels.app.mesh.ble

import android.content.Context
import android.provider.Settings
import dagger.hilt.android.qualifiers.ApplicationContext
import timber.log.Timber
import java.security.MessageDigest
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Generates and persists a unique device identifier for mesh networking.
 *
 * Since Android 6.0 (API 23), BluetoothAdapter.getAddress() returns
 * "02:00:00:00:00:00" for privacy — making MAC-based IDs useless.
 *
 * This class generates a persistent UUID-based ID stored in SharedPreferences,
 * combined with the Android ID for a stable, unique device identity.
 */
@Singleton
class DeviceIdentifier @Inject constructor(
    @ApplicationContext private val context: Context
) {
    companion object {
        private const val TAG = "DeviceIdentifier"
        private const val PREFS_NAME = "ocean_mesh_prefs"
        private const val KEY_DEVICE_ID = "device_mesh_id"
        private const val KEY_DEVICE_FINGERPRINT = "device_fingerprint"
        /** Length of the device ID (hex characters) */
        private const val DEVICE_ID_LENGTH = 12
        /** Length of the fingerprint (hex characters — 16 hex = 8 bytes = 64 bits) */
        private const val FINGERPRINT_LENGTH = 16
    }

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /**
     * Get or generate a persistent unique device ID.
     * Format: 12 hex chars (e.g. "a3f7c1b9e2d4")
     * Used wherever MAC addresses were previously used for device identity.
     */
    fun getDeviceId(): String {
        var id = prefs.getString(KEY_DEVICE_ID, null)
        if (id == null) {
            id = UUID.randomUUID().toString().replace("-", "").take(DEVICE_ID_LENGTH)
            prefs.edit().putString(KEY_DEVICE_ID, id).apply()
            Timber.i("$TAG: Generated new device ID: $id")
        }
        return id
    }

    /**
     * Get or generate a persistent device fingerprint.
     * Combines the persistent device ID with Android ID for extra uniqueness.
     * Format: 16 hex chars (e.g. "a3f7c1b9e2d4f608")
     */
    fun getDeviceFingerprint(): String {
        var fp = prefs.getString(KEY_DEVICE_FINGERPRINT, null)
        if (fp == null) {
            val deviceId = getDeviceId()
            val androidId = try {
                Settings.Secure.getString(
                    context.contentResolver, Settings.Secure.ANDROID_ID
                )
            } catch (e: Exception) {
                Timber.w(e, "$TAG: Failed to read ANDROID_ID")
                null
            } ?: UUID.randomUUID().toString().replace("-", "").take(16)
            fp = generateFingerprint(deviceId, androidId)
            prefs.edit().putString(KEY_DEVICE_FINGERPRINT, fp).apply()
            Timber.i("$TAG: Generated device fingerprint: $fp")
        }
        return fp
    }

    /**
     * Get a mesh ID as a byte array for BLE advertising payloads.
     * Returns the full fingerprint as raw bytes (8 bytes from 16 hex chars).
     */
    fun getMeshIdBytes(): ByteArray {
        val fp = getDeviceFingerprint()
        return hexToByteArray(fp)
    }

    /**
     * Generate a SHA-256 fingerprint from device ID + Android ID.
     * Returns first 16 hex characters (8 bytes / 64 bits).
     */
    private fun generateFingerprint(deviceId: String, androidId: String): String {
        val payload = "$deviceId|$androidId"
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(payload.toByteArray())
        return hash.take(FINGERPRINT_LENGTH / 2).joinToString("") { "%02x".format(it) }
    }

    /**
     * Convert a hex string to a byte array.
     */
    private fun hexToByteArray(hex: String): ByteArray {
        val len = hex.length
        val data = ByteArray(len / 2)
        var i = 0
        while (i < len) {
            data[i / 2] = ((Character.digit(hex[i], 16) shl 4) +
                    Character.digit(hex[i + 1], 16)).toByte()
            i += 2
        }
        return data
    }
}

File: IOS/OceanSentinels/Mesh/BLE/DeviceIdentifier.swift
import Foundation
import CryptoKit
import Security
import os
#if canImport(UIKit)
import UIKit
#endif

/// Generates and persists a unique device identifier for mesh networking.
///
/// On iOS, there's no permanent MAC address access. This class generates
/// a persistent UUID-based ID stored in Keychain (survives app reinstall),
/// combined with identifierForVendor for a stable, unique device identity.
///
/// Replaces Android DeviceIdentifier (SharedPreferences + ANDROID_ID).
final class DeviceIdentifier {
    
    static let shared = DeviceIdentifier()
    
    private let logger = Logger(subsystem: "com.oceansentinels.app", category: "DeviceID")
    
    /// Length of the device ID (hex characters)
    private let deviceIdLength = 12
    /// Length of the fingerprint (hex characters — 16 hex = 8 bytes = 64 bits)
    private let fingerprintLength = 16
    
    // Keychain keys
    private let keychainServiceId = "com.oceansentinels.mesh"
    private let keychainDeviceIdKey = "device_mesh_id"
    private let keychainFingerprintKey = "device_fingerprint"
    
    private init() {}
    
    // MARK: - Device ID
    
    /// Get or generate a persistent unique device ID.
    /// Format: 12 hex chars (e.g. "a3f7c1b9e2d4")
    /// Stored in Keychain to survive app reinstalls.
    func getDeviceId() -> String {
        if let existing = keychainRead(key: keychainDeviceIdKey) {
            return existing
        }
        
        let newId = UUID().uuidString
            .replacingOccurrences(of: "-", with: "")
            .lowercased()
            .prefix(deviceIdLength)
        let idString = String(newId)
        
        keychainWrite(key: keychainDeviceIdKey, value: idString)
        logger.info("Generated new device ID: \(idString)")
        return idString
    }
    
    // MARK: - Device Fingerprint
    
    /// Get or generate a persistent device fingerprint.
    /// Combines the persistent device ID with identifierForVendor.
    /// Format: 16 hex chars (e.g. "a3f7c1b9e2d4f608")
    func getDeviceFingerprint() -> String {
        if let existing = keychainRead(key: keychainFingerprintKey) {
            return existing
        }
        
        let deviceId = getDeviceId()
        
        // Use identifierForVendor as iOS equivalent of ANDROID_ID
        #if os(iOS)
        let vendorId = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
        #else
        let vendorId = UUID().uuidString
        #endif
        
        let fp = generateFingerprint(deviceId: deviceId, vendorId: vendorId)
        keychainWrite(key: keychainFingerprintKey, value: fp)
        logger.info("Generated device fingerprint: \(fp)")
        return fp
    }
    
    /// Get mesh ID as raw bytes for BLE advertising payloads.
    /// Returns 8 bytes from the 16 hex char fingerprint.
    func getMeshIdBytes() -> Data {
        let fp = getDeviceFingerprint()
        return hexToData(fp)
    }
    
    // MARK: - Fingerprint Generation
    
    private func generateFingerprint(deviceId: String, vendorId: String) -> String {
        let payload = "\(deviceId)|\(vendorId)"
        let data = Data(payload.utf8)
        let hash = SHA256.hash(data: data)
        return hash.prefix(fingerprintLength / 2)
            .map { String(format: "%02x", $0) }
            .joined()
    }
    
    // MARK: - Keychain Operations
    
    private func keychainRead(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainServiceId,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess, let data = result as? Data {
            return String(data: data, encoding: .utf8)
        }
        return nil
    }
    
    private func keychainWrite(key: String, value: String) {
        guard let data = value.data(using: .utf8) else { return }
        
        // Delete existing entry first
        let deleteQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainServiceId,
            kSecAttrAccount as String: key
        ]
        SecItemDelete(deleteQuery as CFDictionary)
        
        // Add new entry
        let addQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainServiceId,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock
        ]
        
        let status = SecItemAdd(addQuery as CFDictionary, nil)
        if status != errSecSuccess {
            logger.error("Keychain write failed for \(key): \(status)")
        }
    }
    
    // MARK: - Hex Utility
    
    private func hexToData(_ hex: String) -> Data {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            if let byte = UInt8(hex[index..<nextIndex], radix: 16) {
                data.append(byte)
            }
            index = nextIndex
        }
        return data
    }
}
